lib/spack/spack/test/url_fetch.py

```diff
@@ -358,3 +358,95 @@ def _which(*args, **kwargs):
             with Stage(fetcher, path=testpath) as stage:
                 out = stage.fetch()
             assert err_fmt.format('curl') in out
+
+
+def test_url_fetch_text_without_url(tmpdir):
+    with pytest.raises(web_util.FetchError, match='URL is required'):
+        web_util.fetch_url_text(None)
+
+
+@pytest.mark.parametrize('_fetch_method', ['curl', 'urllib'])
+def test_url_fetch_text_with_invalid_url(tmpdir, _fetch_method):
+    with spack.config.override('config:url_fetch_method', _fetch_method):
+        with pytest.raises(web_util.FetchError):
+            web_util.fetch_url_text('https://no-such-site/')
+
+
+def test_url_missing_curl(tmpdir, monkeypatch):
+    """Check if URL's curl is missing."""
+    url = 'https://github.com'
+    err_fmt = 'No such command {0}'
+
+    def _which(*args, **kwargs):
+        err_msg = err_fmt.format(args[0])
+        raise spack.util.executable.CommandNotFoundError(err_msg)
+
+    # Patching the 'which' symbol imported by spack.util.web needed due
+    # to 'from spack.util.executable import which' in this module.
+    monkeypatch.setattr(spack.util.web, 'which', _which)
+
+    with spack.config.override('config:url_fetch_method', 'curl'):
+        exists = web_util.url_exists(url)
```

Comment:
For some reason I interpreted `default fetch method` to mean it would be used if there BUT, according to you *and* the comments in https://spack.readthedocs.io/en/latest/config_yaml.html, it is not a default but required.  This *should* be changed IMO.

Reply:
> If we require `curl` and it can't be found, IMO that should be an exception (if a user wants to use `web_util.url_exists` in some new context, and it returns `None` because `curl` wasn't available, would they be able to distinguish that from the case that the URL in fact didn't exist?)

Good point.  Should that be part of this PR or a separate one?

Reply:
Removed the check in it duplicates the check in `test_url_fetch_text_curl_failures`.